src/tools/rbd_mirror/ImageSyncThrottler.cc

```diff
@@ -0,0 +1,244 @@
+                            ProgressContext *progress_ctx) {
+  dout(20) << dendl;
+
+  C_SyncHolder *sync_holder_ctx = new C_SyncHolder(this, local_image_ctx->id,
+                                                   on_finish);
+  sync_holder_ctx->m_sync = J<I>::create(local_image_ctx,
+                                         remote_image_ctx, timer,
+                                         timer_lock, mirror_uuid,
+                                         journaler, client_meta,
+                                         work_queue, sync_holder_ctx,
+	                                 progress_ctx);
+  sync_holder_ctx->m_sync->get();
+
+  bool start = false;
+  {
+    Mutex::Locker l(m_lock);
+
+    if (m_inflight_syncs.size() < m_max_concurrent_syncs) {
+      m_inflight_syncs.insert(std::make_pair(local_image_ctx->id,
+                                             sync_holder_ctx));
+      start = true;
+      dout(10) << "ready to start image sync for local_image_id "
+               << local_image_ctx->id << " [" << m_inflight_syncs.size() << "/"
+               << m_max_concurrent_syncs << "]" << dendl;
+    } else {
+      m_sync_queue.push_front(sync_holder_ctx);
+      dout(10) << "image sync for local_image_id " << local_image_ctx->id
+               << " has been queued" << dendl;
+    }
+  }
+
+  if (start) {
+      sync_holder_ctx->m_sync->send();
+  }
+}
+
+template <typename I, template <typename> class J>
+void ImageSyncThrottler<I, J>::cancel_sync(const std::string& local_image_id) {
+  dout(20) << dendl;
+
+  C_SyncHolder *sync_holder = nullptr;
+  bool running_sync = true;
+
+  {
+    Mutex::Locker l(m_lock);
+
+    if (m_inflight_syncs.empty()) {
+      // no image sync currently running and neither waiting
+      return;
+    }
+
+    auto it = m_inflight_syncs.find(local_image_id);
+    if (it != m_inflight_syncs.end()) {
+      sync_holder = it->second;
+    }
+
+    if (!sync_holder) {
+      for (auto it=m_sync_queue.begin(); it != m_sync_queue.end(); ++it) {
+        if ((*it)->m_local_image_id == local_image_id) {
+          sync_holder = (*it);
+          m_sync_queue.erase(it);
+          running_sync = false;
+          break;
+        }
+      }
+    }
+  }
+
+  if (sync_holder) {
+    if (running_sync) {
+      dout(10) << "canceled running image sync for local_image_id "
+               << sync_holder->m_local_image_id << dendl;
+      sync_holder->m_sync->cancel();
+    } else {
+      dout(10) << "canceled waiting image sync for local_image_id "
+               << sync_holder->m_local_image_id << dendl;
+      sync_holder->m_on_finish->complete(-ECANCELED);
+      sync_holder->m_sync->put();
+      delete sync_holder;
+    }
+  }
+}
+
+template <typename I, template <typename> class J>
+void ImageSyncThrottler<I, J>::handle_sync_finished(C_SyncHolder *sync_holder) {
+  dout(20) << dendl;
+
+  C_SyncHolder *next_sync_holder = nullptr;
+
+  {
+    Mutex::Locker l(m_lock);
+
```

Comment:
A call to `put` was already done at this point in the `finish` method of `BaseRequest` class.